### PR TITLE
Adds management of extra inputs and outputs

### PR DESCRIPTION
**Pull request by jeffkole**
Addresses #60 and #62

### CODE REVIEW

**Comment by jeffkole** [comment]
@lisad not complete yet, but it is at a place where your input will be helpful

File: tests/pipelines/employees.py [hunk 1]
```diff
@@ -92,10 +147,34 @@ class Transformer(Phase):
     steps = [
         combine_full_name,
         calculate_annual_salary,
-        calculate_bonus_percent
+        calculate_bonus_percent,
```
Comment (lisad) [inline tests/pipelines/employees.py:150]: I'm hoping to keep the code sample in the readme.md consistent with this actually-useful actually-tested code - not all of it, but I was thinking the phases and pipeline.  But that can wait until this settles

File: tests/test_end_to_end.py [hunk 2]
```diff
@@ -9,18 +9,33 @@
 
 def test_employee_pipeline(tmpdir):
     source = current_path / "fixture_files" / "employees.csv"
-    EmployeeReviewPipeline(source=source, working_dir=tmpdir).run()
+    department_source = current_path / "fixture_files" / "departments.csv"
+    pipeline = EmployeeReviewPipeline(source=source, working_dir=tmpdir)
+    pipeline.init_source('departments', department_source)
```
Comment (lisad) [inline tests/test_end_to_end.py:14]: If it always happens before run, I wonder why it's not part of the constructor.  But I don't feel strongly.

File: tests/pipelines/employees.py [hunk 3]
```diff
@@ -92,10 +147,34 @@ class Transformer(Phase):
...
+        index_departments,
+        add_department_id
+    ]
+    extra_sources = [
+        'departments'
     ]
 
+class ManagerPhase(Phase):
```
Comment (lisad) [inline tests/pipelines/employees.py:158]: Just in order to align with and show an example of some of the common terminology, and assuming this example stays in our front page Readme, I wonder if we should do the phases as:
Enrichment: phase that indexes departments and merges in department id
Transformer: phase that does calculations and reorganizing managers, takes no extra sources

File: phaser/pipeline.py [hunk 4]
```diff
@@ -144,6 +153,41 @@ def __init__(self, working_dir=None, source=None, phases=None):
...
+            return phase.extra_outputs
+        return []
+
+    def init_source(self, name, source):
+        """ Initializes a named source based on the kind of 'source' passed in.
+
+        :param source: must be a os.PathLike file in csv format and will be read entirely into memory
+        """
+        # TODO: Check that file exists first?
+        data = read_csv(source).to_dict('records')
```
Comment (lisad) [inline phaser/pipeline.py:189]: As we gather extra methods that read data, we should allow the Pipeline to have a method for loading data that is not called "read_csv" and can be overridden when somebody customizes the Pipeline.

That can happen later, of course, and even later than that, we will probably have some use cases where 'employees.csv' is loaded as a CSV but 'pay_api_data' is loaded as JSON... then merged...

**Comment by lisad** [comment]
@jeffkole I added a few comments, seems like the right direction

**Comment by jeffkole** [comment]
Goal for the employee pipeline is to have it fit all on one page. 
Add additional pipelines that can show the complexities.
 - File with multiple kinds of events that are emitted as separate streams that are pushed to different phases (employee events -- promotion, pay raise, job change)
 - Traffic incidents or even bicycle and pedestrian counts
 - Retail events -- purchasing services or consumable things

File: phaser/pipeline.py [hunk 5]
```diff
@@ -73,6 +74,14 @@ def add_output(self, name, data):
         # At present outputs must be in record format and save to CSV, but this should be expanded.
         self.outputs.append(ReadWriteObject(name, data, to_save=True))
 
+    def set_source(self, name, data):
+        self.sources[name] = data
```
Comment (lisad) [inline phaser/pipeline.py:78]: Should sources be ReadWriteObjects too?

File: tests/test_multi_source_and_outputs.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import pandas
+from pathlib import Path
+from pipelines.multi_source_and_outputs import EmployeeReviewPipeline
+from phaser.io import read_csv
+
+current_path = Path(__file__).parent
+
+
+def test_employee_pipeline(tmpdir):
+    print(tmpdir)
```
Comment (lisad) [inline tests/test_multi_source_and_outputs.py:14]: Stray prints here and below?

**Review by lisad — APPROVED** [review]

**Pull request merged**